tags/Generic/TaxonomiesTracker/Taxonomy Tracker.tpl

```diff
@@ -0,0 +1,138 @@
+    "displayName": "",
+    "id": "brand_dummy"
+  },
+  "containerContexts": [
+    "WEB"
+  ]
+}
+
+
+___TEMPLATE_PARAMETERS___
+
+[
+  {
+    "valueValidators": [
+      {
+        "type": "NON_EMPTY"
+      }
+    ],
+    "displayName": "Taxonomy Name",
+    "simpleValueType": true,
+    "name": "taxonomyName",
+    "type": "TEXT"
+  },
+  {
+    "macrosInSelect": true,
+    "selectItems": [],
+    "valueValidators": [
+      {
+        "type": "NON_EMPTY"
+      },
+      {
+        "args": [],
+        "type": "TABLE_ROW_COUNT"
+      }
+    ],
+    "displayName": "Taxonomy Values Array",
+    "simpleValueType": true,
+    "name": "taxonomyValuesArray",
+    "type": "SELECT"
+  }
+]
+
+
+___WEB_PERMISSIONS___
+
+[
+  {
+    "instance": {
+      "key": {
+        "publicId": "access_globals",
+        "versionId": "1"
+      },
+      "param": [
+        {
+          "key": "keys",
+          "value": {
+            "type": 2,
+            "listItem": [
+              {
+                "type": 3,
+                "mapKey": [
+                  {
+                    "type": 1,
+                    "string": "key"
+                  },
+                  {
+                    "type": 1,
+                    "string": "read"
+                  },
+                  {
+                    "type": 1,
+                    "string": "write"
+                  },
+                  {
+                    "type": 1,
+                    "string": "execute"
+                  }
+                ],
+                "mapValue": [
+                  {
+                    "type": 1,
+                    "string": "dataLayer"
+                  },
+                  {
+                    "type": 8,
+                    "boolean": true
+                  },
+                  {
+                    "type": 8,
+                    "boolean": true
+                  },
+                  {
+                    "type": 8,
+                    "boolean": false
+                  }
+                ]
+              }
+            ]
+          }
+        }
+      ]
+    },
+    "clientAnnotations": {
+      "isEditedByUser": true
+    },
+    "isRequired": true
+  }
+]
+
+
+___SANDBOXED_JS_FOR_WEB_TEMPLATE___
+
+const createQueue = require('createQueue');
+const dataLayerPush = createQueue('dataLayer');
+const taxonomyValuesArray = data.taxonomyValuesArray;
+
+for(var i = 0; i<taxonomyValuesArray.length; i++){
+  dataLayerPush({'taxonomyValue': taxonomyValuesArray[i]}, {'taxonomyName': data.taxonomyName},{'event': 'taxonomyTracker'});
+}
+
```

Comment:
Currently you are pushing **three** different objects with each iteration. One push will have "taxonomyValue", one will have "taxonomyName" and one will have "event". That's super ineffective. Here's how I would refactor it:

```
// Check if taxonomyValues is an array to avoid an error
const isArray = arr => callInWindow('toString.call', arr) === '[object Array]';

if (isArray(data.taxonomyValuesArray)) {
  data.taxonomyValuesArray.forEach(t => dataLayerPush({
    taxonomyValue: t,
    taxonomyName: data.taxonomyName,
    event: 'taxonomyTracker'
  }));
}
```

Note you'll also need to add `toString.call` with "Execute" rights as a permission.

You can do without the check, but in that case any item other than a variable will throw an error.

At the very least, the items should be combined into a single object with three keys rather than three objects with one key each.

Reply:
Good to know. That's what I thought originally but I was confused by Google's documentation.

https://developers.google.com/tag-manager/templates/api#createqueue

Anyway, I've fixed it.